### PR TITLE
Some refactor on git diff and ignore getting commit information failed on migrating pull request review comments

### PR DESCRIPTION
**Pull request by lunny**
When migrating a repository from github with pull requests, if the pull request review comments refs an outdate commit which is not yet on this repository any more. We should ignore that commits.

This pull request also moved some git diff functions from package `gitdiff` to `git`.

### CODE REVIEW

**Review by 6543 — APPROVED** [review]

**Comment by 6543** [comment]
Beside **modules/migrations/gitea.go** it is a refactor

**Comment by zeripath** [comment]
@lunny it's hard to see what you've fixed and what's been moved - I wonder could you split the commits in two - one doing the move and one doing the fix?

File: modules/git/diff.go [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+		leftHunk, _ = strconv.Atoi(leftRange[1])
+	}
+	if len(ranges) > 1 {
+		rightRange := strings.Split(ranges[1], ",")
+		rightLine, _ = strconv.Atoi(rightRange[0])
+		if len(rightRange) > 1 {
+			righHunk, _ = strconv.Atoi(rightRange[1])
+		}
+	} else {
+		log("Parse line number failed: %v", diffhunk)
```
Comment (guillep2k) [inline modules/git/diff.go:116]: Shouldn't this go to the regular log? (in fact, I wonder why git has it's own log I can't redirect...)

File: modules/migrations/gitea.go [hunk 2]
```diff
@@ -783,19 +782,22 @@ func (g *GiteaLocalUploader) CreateReviews(reviews ...*base.Review) error {
...
 				return fmt.Errorf("GetRefCommitID[%s]: %v", pr.GetGitRefName(), err)
 			}
+
+			var patch string
 			patchBuf := new(bytes.Buffer)
-			if err := gitdiff.GetRawDiffForFile(g.gitRepo.Path, pr.MergeBase, headCommitID, gitdiff.RawDiffNormal, comment.TreePath, patchBuf); err != nil {
-				return fmt.Errorf("GetRawDiffForLine[%s, %s, %s, %s]: %v", g.gitRepo.Path, pr.MergeBase, headCommitID, comment.TreePath, err)
+			if err := git.GetRepoRawDiffForFile(g.gitRepo, pr.MergeBase, headCommitID, git.RawDiffNormal, comment.TreePath, patchBuf); err != nil {
+				// We should ignore the error since the commit maybe removed when force push to the pull request
+				log.Error("GetRepoRawDiffForFile failed when migrating [%s, %s, %s, %s]: %v", g.gitRepo.Path, pr.MergeBase, headCommitID, comment.TreePath, err)
```
Comment (guillep2k) [inline modules/migrations/gitea.go]:
```suggestion
				log.Warn("GetRepoRawDiffForFile failed when migrating [%s, %s, %s, %s]: %v", g.gitRepo.Path, pr.MergeBase, headCommitID, comment.TreePath, err)
```

**Review by guillep2k — APPROVED** [review]
Just a couple of nits.

File: modules/git/diff.go (hunk 1 above)
Comment (lunny) [reply modules/git/diff.go:116]: package `git` maybe invoked via `gitea serv` and `gitea hooks` which will output information to SSH client. So git library has it's own log.

**Pull request merged**